lib/compilers/coccinelle.ts

```diff
@@ -0,0 +1,281 @@
+        return super.getIrOutputFilename(inputFilename, filters);
+    }
+
+    override getArgumentParserClass() {
+        return super.getArgumentParserClass();
+    }
+
+    override isCfgCompiler() {
+        return false;
+    }
+
+    override processExecutionResult(input: UnprocessedExecResult, inputFilename?: string): BasicExecutionResult {
+        return utils.processExecutionResult(input, inputFilename);
+    }
+
+    override buildExecutable(
+        compiler: string,
+        options: string[],
+        inputFilename: string,
+        execOptions: ExecutionOptionsWithEnv,
+    ) {
+        // may move here the file separation logic currently in runCompiler
+        return super.buildExecutable(compiler, options, inputFilename, execOptions);
+    }
+
+    checkCocciSource(source: string) {
+        // forbid '#spatch'-alike lines anywhere in the semantic patch
+        const re = /^\s*#\s*spatch.*/;
+        const failed: string[] = [];
+        for (const [index, line] of utils.splitLines(source).entries()) {
+            if (re.test(line)) {
+                failed.push(`<stdin>:${index + 1}:1: no #spatch lines please`);
+            }
+        }
+        if (failed.length > 0) return failed.join('\n');
+        return null;
+    }
+
+    override async runCompiler(
+        compiler: string,
+        options: string[],
+        inputFilename: string,
+        execOptions: ExecutionOptionsWithEnv,
+        filters?: ParseFiltersAndOutputOptions,
+    ): Promise<CompilationResult> {
+        // here to extract contents of source and semantic patch
+        if (!execOptions) {
+            execOptions = this.getDefaultExecOptions();
+        }
+
+        if (!execOptions.customCwd) {
+            execOptions.customCwd = path.dirname(inputFilename);
+        }
+
+        const spatchFilename = path.join(path.dirname(inputFilename), this.spatchBaseFilename);
+        let cFileContents = '';
+        let pFileContents = '';
+        let toc = true; // toc: Target Language or Coccinelle Language
+        const file = await open(inputFilename);
+        for await (const line of file.readLines()) {
+            // this separates the C/C++ portion from the SmPL
+            const ifdefCocciRegex = /^#ifdef.*COCCINELLE.*/;
+            const endifCocciRegex = /^#endif.*COCCINELLE.*/;
+
+            if (ifdefCocciRegex.test(line) && toc) {
+                toc = !toc;
+            } else {
+                if (!endifCocciRegex.test(line)) {
+                    if (toc) cFileContents += line + '\n';
+                    else pFileContents += line + '\n';
+                }
+            }
+        }
+        file.close();
+
+        // we overwrite the C/C++ + SmPL with extracted C/C++ sources
+        fs.writeFileSync(inputFilename, cFileContents);
+        // we save the extracted SmPL sources
+        fs.writeFileSync(spatchFilename, pFileContents);
```

Comment:
Please use node:fs/promises here to avoid blocking the event loop

Reply:
Done :+1: -- please see https://github.com/compiler-explorer/compiler-explorer/pull/7001/commits/a08224883761af121f846d0b5c9d49df2268292e

Reply:
Great, thanks